{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Exploratory Data Analysis of Cancer Genomics data using TCGA\n",
    "\n",
    "In this notebook, we will take a look at one of the canonical datasets, if not _the_ dataset, in cancer genomics: TCGA.\n",
    "\n",
    "We'll start with investigating the RNA Sequencing (rnaseq) and Clinical data available for a type of liver cancer known as hepatocellular carcinoma (HCC). Hepatocellular carcinoma is the most common form of liver cancer in the United States, making up [more than 80% of cases](https://www.cancer.gov/about-nci/organization/ccg/research/structural-genomics/tcga/studied-cancers/liver). The TCGA dataset is abbreviated LIHC.\n",
    "\n",
    "Some examples of what researchers have learned from the LIHC dataset at the DNA-level include confirmed [frequent mutations](https://www.cancer.gov/about-nci/organization/ccg/research/structural-genomics/tcga/studied-cancers/liver) in:\n",
    " - The TERT promotor region, associated with regulating cell survival\n",
    " - TP53, one of the most frequently mutated genes in cancer\n",
    " - CTNNB1, a member of the Wnt signaling pathway that mediates cell growth and differentiation\n",
    "     \n",
    "There are currently several therapies under development that target these genes.\n",
    "\n",
    "In addition to DNA alterations however, different biological and tumor microenvrionment factors can [influence disease progression](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6237857/). A transcriptomic survey of tissues at various stages of disease progression could help elucidate some of the underlying pathways contributing to tumorigenesis.   \n",
    "\n",
    "### Today, we'll be focusing on using RNA-seq data from LIHC combined with clinical attributes to identify biomarkers for disease progression. \n",
    " \n",
    "The data is stored in the R package _[RTCGA](http://rtcga.github.io/RTCGA/)_"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "heading_collapsed": true
   },
   "source": [
    "## Load libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "heading_collapsed": true
   },
   "source": [
    "## Set variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true,
    "hidden": true
   },
   "outputs": [],
   "source": [
    "data_dir=\"data/\"\n",
    "response_name=\"patient.race\"\n",
    "rnaseq_file=data_dir+\"lihc_rnaseq.csv.gz\"\n",
    "clinical_file=data_dir+\"lihc_clinical.csv.gz\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "heading_collapsed": true
   },
   "source": [
    "## Load data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "hidden": true
   },
   "source": [
    "The data is stored in the RTCGA package in the R programming language. I've outputted it for easy use within python. \n",
    "\n",
    "We will be investigating the Hepatocellular carcinoma dataset. Read about it [here](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5680778/).\n",
    "\n",
    "The TCGA RNASeq data is illumina hiseq Level 3 RSEM normalized expression data. You can read about thec RSEM method [here](https://academic.oup.com/bioinformatics/article/26/4/493/243395).\n",
    "\n",
    "Essentially this is the raw counts of reads that aligned to the gene transcript, though it's only a guess by the program. Since it's a guess, the values are rational numbers. To simplify things, we'll round the values to the next whole integer. \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "hidden": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(423, 20531)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>?|100130426</th>\n",
       "      <th>?|100133144</th>\n",
       "      <th>?|100134869</th>\n",
       "      <th>?|10357</th>\n",
       "      <th>?|10431</th>\n",
       "      <th>?|136542</th>\n",
       "      <th>?|155060</th>\n",
       "      <th>?|26823</th>\n",
       "      <th>?|280660</th>\n",
       "      <th>?|317712</th>\n",
       "      <th>...</th>\n",
       "      <th>ZXDA|7789</th>\n",
       "      <th>ZXDB|158586</th>\n",
       "      <th>ZXDC|79364</th>\n",
       "      <th>ZYG11A|440590</th>\n",
       "      <th>ZYG11B|79699</th>\n",
       "      <th>ZYX|7791</th>\n",
       "      <th>ZZEF1|23140</th>\n",
       "      <th>ZZZ3|26009</th>\n",
       "      <th>psiTPTE22|387590</th>\n",
       "      <th>tAKR|389932</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>bcr_patient_barcode</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>TCGA-2V-A95S-01A-11R-A37K-07</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>91</td>\n",
       "      <td>1018</td>\n",
       "      <td>0</td>\n",
       "      <td>142</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>25</td>\n",
       "      <td>274</td>\n",
       "      <td>795</td>\n",
       "      <td>19</td>\n",
       "      <td>500</td>\n",
       "      <td>3173</td>\n",
       "      <td>891</td>\n",
       "      <td>511</td>\n",
       "      <td>4</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>TCGA-2Y-A9GS-01A-12R-A38B-07</td>\n",
       "      <td>0</td>\n",
       "      <td>27</td>\n",
       "      <td>3</td>\n",
       "      <td>72</td>\n",
       "      <td>640</td>\n",
       "      <td>0</td>\n",
       "      <td>123</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>69</td>\n",
       "      <td>633</td>\n",
       "      <td>1154</td>\n",
       "      <td>72</td>\n",
       "      <td>1001</td>\n",
       "      <td>5302</td>\n",
       "      <td>756</td>\n",
       "      <td>861</td>\n",
       "      <td>7</td>\n",
       "      <td>483</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>TCGA-2Y-A9GT-01A-11R-A38B-07</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>96</td>\n",
       "      <td>743</td>\n",
       "      <td>0</td>\n",
       "      <td>96</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>47</td>\n",
       "      <td>1220</td>\n",
       "      <td>1134</td>\n",
       "      <td>13</td>\n",
       "      <td>1290</td>\n",
       "      <td>3220</td>\n",
       "      <td>861</td>\n",
       "      <td>524</td>\n",
       "      <td>15</td>\n",
       "      <td>84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>TCGA-2Y-A9GU-01A-11R-A38B-07</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>62</td>\n",
       "      <td>1187</td>\n",
       "      <td>0</td>\n",
       "      <td>281</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>19</td>\n",
       "      <td>286</td>\n",
       "      <td>1151</td>\n",
       "      <td>10</td>\n",
       "      <td>942</td>\n",
       "      <td>3093</td>\n",
       "      <td>1340</td>\n",
       "      <td>344</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>TCGA-2Y-A9GV-01A-11R-A38B-07</td>\n",
       "      <td>0</td>\n",
       "      <td>12</td>\n",
       "      <td>6</td>\n",
       "      <td>105</td>\n",
       "      <td>879</td>\n",
       "      <td>0</td>\n",
       "      <td>283</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>42</td>\n",
       "      <td>1000</td>\n",
       "      <td>1632</td>\n",
       "      <td>5</td>\n",
       "      <td>1381</td>\n",
       "      <td>2903</td>\n",
       "      <td>576</td>\n",
       "      <td>666</td>\n",
       "      <td>3</td>\n",
       "      <td>120</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 20531 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                              ?|100130426  ?|100133144  ?|100134869  ?|10357  \\\n",
       "bcr_patient_barcode                                                            \n",
       "TCGA-2V-A95S-01A-11R-A37K-07            0            2            4       91   \n",
       "TCGA-2Y-A9GS-01A-12R-A38B-07            0           27            3       72   \n",
       "TCGA-2Y-A9GT-01A-11R-A38B-07            0            0            5       96   \n",
       "TCGA-2Y-A9GU-01A-11R-A38B-07            0            6            6       62   \n",
       "TCGA-2Y-A9GV-01A-11R-A38B-07            0           12            6      105   \n",
       "\n",
       "                              ?|10431  ?|136542  ?|155060  ?|26823  ?|280660  \\\n",
       "bcr_patient_barcode                                                            \n",
       "TCGA-2V-A95S-01A-11R-A37K-07     1018         0       142        1         0   \n",
       "TCGA-2Y-A9GS-01A-12R-A38B-07      640         0       123        2         0   \n",
       "TCGA-2Y-A9GT-01A-11R-A38B-07      743         0        96        2         1   \n",
       "TCGA-2Y-A9GU-01A-11R-A38B-07     1187         0       281        1         0   \n",
       "TCGA-2Y-A9GV-01A-11R-A38B-07      879         0       283        0         0   \n",
       "\n",
       "                              ?|317712  ...  ZXDA|7789  ZXDB|158586  \\\n",
       "bcr_patient_barcode                     ...                           \n",
       "TCGA-2V-A95S-01A-11R-A37K-07         0  ...         25          274   \n",
       "TCGA-2Y-A9GS-01A-12R-A38B-07         0  ...         69          633   \n",
       "TCGA-2Y-A9GT-01A-11R-A38B-07         0  ...         47         1220   \n",
       "TCGA-2Y-A9GU-01A-11R-A38B-07         0  ...         19          286   \n",
       "TCGA-2Y-A9GV-01A-11R-A38B-07         0  ...         42         1000   \n",
       "\n",
       "                              ZXDC|79364  ZYG11A|440590  ZYG11B|79699  \\\n",
       "bcr_patient_barcode                                                     \n",
       "TCGA-2V-A95S-01A-11R-A37K-07         795             19           500   \n",
       "TCGA-2Y-A9GS-01A-12R-A38B-07        1154             72          1001   \n",
       "TCGA-2Y-A9GT-01A-11R-A38B-07        1134             13          1290   \n",
       "TCGA-2Y-A9GU-01A-11R-A38B-07        1151             10           942   \n",
       "TCGA-2Y-A9GV-01A-11R-A38B-07        1632              5          1381   \n",
       "\n",
       "                              ZYX|7791  ZZEF1|23140  ZZZ3|26009  \\\n",
       "bcr_patient_barcode                                               \n",
       "TCGA-2V-A95S-01A-11R-A37K-07      3173          891         511   \n",
       "TCGA-2Y-A9GS-01A-12R-A38B-07      5302          756         861   \n",
       "TCGA-2Y-A9GT-01A-11R-A38B-07      3220          861         524   \n",
       "TCGA-2Y-A9GU-01A-11R-A38B-07      3093         1340         344   \n",
       "TCGA-2Y-A9GV-01A-11R-A38B-07      2903          576         666   \n",
       "\n",
       "                              psiTPTE22|387590  tAKR|389932  \n",
       "bcr_patient_barcode                                          \n",
       "TCGA-2V-A95S-01A-11R-A37K-07                 4            7  \n",
       "TCGA-2Y-A9GS-01A-12R-A38B-07                 7          483  \n",
       "TCGA-2Y-A9GT-01A-11R-A38B-07                15           84  \n",
       "TCGA-2Y-A9GU-01A-11R-A38B-07                 3            3  \n",
       "TCGA-2Y-A9GV-01A-11R-A38B-07                 3          120  \n",
       "\n",
       "[5 rows x 20531 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "rnaseq = (pd.\n",
    "          read_csv(rnaseq_file,compression=\"gzip\").\n",
    "          set_index('bcr_patient_barcode').\n",
    "          applymap(lambda x : int(np.ceil(x)))\n",
    "         )\n",
    "display(rnaseq.shape)\n",
    "display(rnaseq.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "hidden": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2BP1</th>\n",
       "      <th>A2LD1</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>A4GNT</th>\n",
       "      <th>AAA1</th>\n",
       "      <th>AAAS</th>\n",
       "      <th>...</th>\n",
       "      <th>ZXDA</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "      <th>psiTPTE22</th>\n",
       "      <th>tAKR</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>bcr_patient_barcode</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>TCGA-2V-A95S-01A-11R-A37K-07</td>\n",
       "      <td>22283</td>\n",
       "      <td>584</td>\n",
       "      <td>0</td>\n",
       "      <td>375</td>\n",
       "      <td>0</td>\n",
       "      <td>286320</td>\n",
       "      <td>81</td>\n",
       "      <td>543</td>\n",
       "      <td>1</td>\n",
       "      <td>1032</td>\n",
       "      <td>...</td>\n",
       "      <td>25</td>\n",
       "      <td>274</td>\n",
       "      <td>795</td>\n",
       "      <td>19</td>\n",
       "      <td>500</td>\n",
       "      <td>3173</td>\n",
       "      <td>891</td>\n",
       "      <td>511</td>\n",
       "      <td>4</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>TCGA-2Y-A9GS-01A-12R-A38B-07</td>\n",
       "      <td>22642</td>\n",
       "      <td>1573</td>\n",
       "      <td>3</td>\n",
       "      <td>99</td>\n",
       "      <td>0</td>\n",
       "      <td>31169</td>\n",
       "      <td>163</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>903</td>\n",
       "      <td>...</td>\n",
       "      <td>69</td>\n",
       "      <td>633</td>\n",
       "      <td>1154</td>\n",
       "      <td>72</td>\n",
       "      <td>1001</td>\n",
       "      <td>5302</td>\n",
       "      <td>756</td>\n",
       "      <td>861</td>\n",
       "      <td>7</td>\n",
       "      <td>483</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>TCGA-2Y-A9GT-01A-11R-A38B-07</td>\n",
       "      <td>77670</td>\n",
       "      <td>1281</td>\n",
       "      <td>0</td>\n",
       "      <td>215</td>\n",
       "      <td>1</td>\n",
       "      <td>19515</td>\n",
       "      <td>119</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>773</td>\n",
       "      <td>...</td>\n",
       "      <td>47</td>\n",
       "      <td>1220</td>\n",
       "      <td>1134</td>\n",
       "      <td>13</td>\n",
       "      <td>1290</td>\n",
       "      <td>3220</td>\n",
       "      <td>861</td>\n",
       "      <td>524</td>\n",
       "      <td>15</td>\n",
       "      <td>84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>TCGA-2Y-A9GU-01A-11R-A38B-07</td>\n",
       "      <td>9323</td>\n",
       "      <td>1253</td>\n",
       "      <td>0</td>\n",
       "      <td>2914</td>\n",
       "      <td>2</td>\n",
       "      <td>243941</td>\n",
       "      <td>72</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>722</td>\n",
       "      <td>...</td>\n",
       "      <td>19</td>\n",
       "      <td>286</td>\n",
       "      <td>1151</td>\n",
       "      <td>10</td>\n",
       "      <td>942</td>\n",
       "      <td>3093</td>\n",
       "      <td>1340</td>\n",
       "      <td>344</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>TCGA-2Y-A9GV-01A-11R-A38B-07</td>\n",
       "      <td>84243</td>\n",
       "      <td>1641</td>\n",
       "      <td>0</td>\n",
       "      <td>404</td>\n",
       "      <td>0</td>\n",
       "      <td>8756</td>\n",
       "      <td>83</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>828</td>\n",
       "      <td>...</td>\n",
       "      <td>42</td>\n",
       "      <td>1000</td>\n",
       "      <td>1632</td>\n",
       "      <td>5</td>\n",
       "      <td>1381</td>\n",
       "      <td>2903</td>\n",
       "      <td>576</td>\n",
       "      <td>666</td>\n",
       "      <td>3</td>\n",
       "      <td>120</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 20501 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                               A1BG  A1CF  A2BP1  A2LD1  A2ML1     A2M  \\\n",
       "bcr_patient_barcode                                                      \n",
       "TCGA-2V-A95S-01A-11R-A37K-07  22283   584      0    375      0  286320   \n",
       "TCGA-2Y-A9GS-01A-12R-A38B-07  22642  1573      3     99      0   31169   \n",
       "TCGA-2Y-A9GT-01A-11R-A38B-07  77670  1281      0    215      1   19515   \n",
       "TCGA-2Y-A9GU-01A-11R-A38B-07   9323  1253      0   2914      2  243941   \n",
       "TCGA-2Y-A9GV-01A-11R-A38B-07  84243  1641      0    404      0    8756   \n",
       "\n",
       "                              A4GALT  A4GNT  AAA1  AAAS  ...  ZXDA  ZXDB  \\\n",
       "bcr_patient_barcode                                      ...               \n",
       "TCGA-2V-A95S-01A-11R-A37K-07      81    543     1  1032  ...    25   274   \n",
       "TCGA-2Y-A9GS-01A-12R-A38B-07     163      2     1   903  ...    69   633   \n",
       "TCGA-2Y-A9GT-01A-11R-A38B-07     119      1     3   773  ...    47  1220   \n",
       "TCGA-2Y-A9GU-01A-11R-A38B-07      72      0     0   722  ...    19   286   \n",
       "TCGA-2Y-A9GV-01A-11R-A38B-07      83      0     9   828  ...    42  1000   \n",
       "\n",
       "                              ZXDC  ZYG11A  ZYG11B   ZYX  ZZEF1  ZZZ3  \\\n",
       "bcr_patient_barcode                                                     \n",
       "TCGA-2V-A95S-01A-11R-A37K-07   795      19     500  3173    891   511   \n",
       "TCGA-2Y-A9GS-01A-12R-A38B-07  1154      72    1001  5302    756   861   \n",
       "TCGA-2Y-A9GT-01A-11R-A38B-07  1134      13    1290  3220    861   524   \n",
       "TCGA-2Y-A9GU-01A-11R-A38B-07  1151      10     942  3093   1340   344   \n",
       "TCGA-2Y-A9GV-01A-11R-A38B-07  1632       5    1381  2903    576   666   \n",
       "\n",
       "                              psiTPTE22  tAKR  \n",
       "bcr_patient_barcode                            \n",
       "TCGA-2V-A95S-01A-11R-A37K-07          4     7  \n",
       "TCGA-2Y-A9GS-01A-12R-A38B-07          7   483  \n",
       "TCGA-2Y-A9GT-01A-11R-A38B-07         15    84  \n",
       "TCGA-2Y-A9GU-01A-11R-A38B-07          3     3  \n",
       "TCGA-2Y-A9GV-01A-11R-A38B-07          3   120  \n",
       "\n",
       "[5 rows x 20501 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_name_logical = [len(x[0])>1 for x in rnaseq.columns.str.split('|')]\n",
    "sub = rnaseq.loc[:,gene_name_logical]\n",
    "sub.columns = [x[0] for x in sub.columns.str.split('|')]\n",
    "rnaseq_sub = sub.copy()\n",
    "rnaseq_sub.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "hidden": true
   },
   "source": [
    "The clinical data is within the RTCGA package, but is also available [here](https://portal.gdc.cancer.gov/projects/TCGA-LIHC). More cdescription of the clinical attributes are [here](https://gdc.cancer.gov/about-data/data-harmonization-and-generation/clinical-data-harmonization)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "hidden": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(377, 1586)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>admin.bcr</th>\n",
       "      <th>admin.day_of_dcc_upload</th>\n",
       "      <th>admin.disease_code</th>\n",
       "      <th>admin.file_uuid</th>\n",
       "      <th>admin.month_of_dcc_upload</th>\n",
       "      <th>admin.patient_withdrawal.withdrawn</th>\n",
       "      <th>admin.project_code</th>\n",
       "      <th>admin.year_of_dcc_upload</th>\n",
       "      <th>patient.ablation_embolization_tx_adjuvant</th>\n",
       "      <th>patient.additional_studies</th>\n",
       "      <th>...</th>\n",
       "      <th>patient.samples.sample.preservation_method</th>\n",
       "      <th>patient.samples.sample.sample_type</th>\n",
       "      <th>patient.samples.sample.sample_type_id</th>\n",
       "      <th>patient.samples.sample.shortest_dimension</th>\n",
       "      <th>patient.samples.sample.time_between_clamping_and_freezing</th>\n",
       "      <th>patient.samples.sample.time_between_excision_and_freezing</th>\n",
       "      <th>patient.samples.sample.tissue_type</th>\n",
       "      <th>patient.samples.sample.tumor_descriptor</th>\n",
       "      <th>patient.samples.sample.tumor_pathology</th>\n",
       "      <th>patient.samples.sample.vial_number</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>patient.bcr_patient_barcode</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>tcga-2v-a95s</td>\n",
       "      <td>nationwide children's hospital</td>\n",
       "      <td>1</td>\n",
       "      <td>lihc</td>\n",
       "      <td>47a9da83-f87b-45a0-a501-542c5a9df212</td>\n",
       "      <td>10</td>\n",
       "      <td>False</td>\n",
       "      <td>tcga</td>\n",
       "      <td>2015</td>\n",
       "      <td>no</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>primary tumor</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>tcga-2y-a9gs</td>\n",
       "      <td>nationwide children's hospital</td>\n",
       "      <td>1</td>\n",
       "      <td>lihc</td>\n",
       "      <td>ada84bc2-9724-428c-9557-7280af5f8297</td>\n",
       "      <td>10</td>\n",
       "      <td>False</td>\n",
       "      <td>tcga</td>\n",
       "      <td>2015</td>\n",
       "      <td>no</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>primary tumor</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>tcga-2y-a9gt</td>\n",
       "      <td>nationwide children's hospital</td>\n",
       "      <td>1</td>\n",
       "      <td>lihc</td>\n",
       "      <td>b9f63d47-2293-4360-b373-5acdaa2ec7cb</td>\n",
       "      <td>10</td>\n",
       "      <td>False</td>\n",
       "      <td>tcga</td>\n",
       "      <td>2015</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>primary tumor</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>tcga-2y-a9gu</td>\n",
       "      <td>nationwide children's hospital</td>\n",
       "      <td>1</td>\n",
       "      <td>lihc</td>\n",
       "      <td>2b488ba2-f46b-4d0b-8c73-d9a761ab61b1</td>\n",
       "      <td>10</td>\n",
       "      <td>False</td>\n",
       "      <td>tcga</td>\n",
       "      <td>2015</td>\n",
       "      <td>no</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>primary tumor</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>tcga-2y-a9gv</td>\n",
       "      <td>nationwide children's hospital</td>\n",
       "      <td>1</td>\n",
       "      <td>lihc</td>\n",
       "      <td>69fc70f4-fe6a-472a-b354-cf37515c524d</td>\n",
       "      <td>10</td>\n",
       "      <td>False</td>\n",
       "      <td>tcga</td>\n",
       "      <td>2015</td>\n",
       "      <td>no</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>primary tumor</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>a</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 1586 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                  admin.bcr  \\\n",
       "patient.bcr_patient_barcode                                   \n",
       "tcga-2v-a95s                 nationwide children's hospital   \n",
       "tcga-2y-a9gs                 nationwide children's hospital   \n",
       "tcga-2y-a9gt                 nationwide children's hospital   \n",
       "tcga-2y-a9gu                 nationwide children's hospital   \n",
       "tcga-2y-a9gv                 nationwide children's hospital   \n",
       "\n",
       "                             admin.day_of_dcc_upload admin.disease_code  \\\n",
       "patient.bcr_patient_barcode                                               \n",
       "tcga-2v-a95s                                       1               lihc   \n",
       "tcga-2y-a9gs                                       1               lihc   \n",
       "tcga-2y-a9gt                                       1               lihc   \n",
       "tcga-2y-a9gu                                       1               lihc   \n",
       "tcga-2y-a9gv                                       1               lihc   \n",
       "\n",
       "                                                  admin.file_uuid  \\\n",
       "patient.bcr_patient_barcode                                         \n",
       "tcga-2v-a95s                 47a9da83-f87b-45a0-a501-542c5a9df212   \n",
       "tcga-2y-a9gs                 ada84bc2-9724-428c-9557-7280af5f8297   \n",
       "tcga-2y-a9gt                 b9f63d47-2293-4360-b373-5acdaa2ec7cb   \n",
       "tcga-2y-a9gu                 2b488ba2-f46b-4d0b-8c73-d9a761ab61b1   \n",
       "tcga-2y-a9gv                 69fc70f4-fe6a-472a-b354-cf37515c524d   \n",
       "\n",
       "                             admin.month_of_dcc_upload  \\\n",
       "patient.bcr_patient_barcode                              \n",
       "tcga-2v-a95s                                        10   \n",
       "tcga-2y-a9gs                                        10   \n",
       "tcga-2y-a9gt                                        10   \n",
       "tcga-2y-a9gu                                        10   \n",
       "tcga-2y-a9gv                                        10   \n",
       "\n",
       "                             admin.patient_withdrawal.withdrawn  \\\n",
       "patient.bcr_patient_barcode                                       \n",
       "tcga-2v-a95s                                              False   \n",
       "tcga-2y-a9gs                                              False   \n",
       "tcga-2y-a9gt                                              False   \n",
       "tcga-2y-a9gu                                              False   \n",
       "tcga-2y-a9gv                                              False   \n",
       "\n",
       "                            admin.project_code  admin.year_of_dcc_upload  \\\n",
       "patient.bcr_patient_barcode                                                \n",
       "tcga-2v-a95s                              tcga                      2015   \n",
       "tcga-2y-a9gs                              tcga                      2015   \n",
       "tcga-2y-a9gt                              tcga                      2015   \n",
       "tcga-2y-a9gu                              tcga                      2015   \n",
       "tcga-2y-a9gv                              tcga                      2015   \n",
       "\n",
       "                            patient.ablation_embolization_tx_adjuvant  \\\n",
       "patient.bcr_patient_barcode                                             \n",
       "tcga-2v-a95s                                                       no   \n",
       "tcga-2y-a9gs                                                       no   \n",
       "tcga-2y-a9gt                                                      NaN   \n",
       "tcga-2y-a9gu                                                       no   \n",
       "tcga-2y-a9gv                                                       no   \n",
       "\n",
       "                             patient.additional_studies  ...  \\\n",
       "patient.bcr_patient_barcode                              ...   \n",
       "tcga-2v-a95s                                        NaN  ...   \n",
       "tcga-2y-a9gs                                        NaN  ...   \n",
       "tcga-2y-a9gt                                        NaN  ...   \n",
       "tcga-2y-a9gu                                        NaN  ...   \n",
       "tcga-2y-a9gv                                        NaN  ...   \n",
       "\n",
       "                            patient.samples.sample.preservation_method  \\\n",
       "patient.bcr_patient_barcode                                              \n",
       "tcga-2v-a95s                                                       NaN   \n",
       "tcga-2y-a9gs                                                       NaN   \n",
       "tcga-2y-a9gt                                                       NaN   \n",
       "tcga-2y-a9gu                                                       NaN   \n",
       "tcga-2y-a9gv                                                       NaN   \n",
       "\n",
       "                             patient.samples.sample.sample_type  \\\n",
       "patient.bcr_patient_barcode                                       \n",
       "tcga-2v-a95s                                      primary tumor   \n",
       "tcga-2y-a9gs                                      primary tumor   \n",
       "tcga-2y-a9gt                                      primary tumor   \n",
       "tcga-2y-a9gu                                      primary tumor   \n",
       "tcga-2y-a9gv                                      primary tumor   \n",
       "\n",
       "                            patient.samples.sample.sample_type_id  \\\n",
       "patient.bcr_patient_barcode                                         \n",
       "tcga-2v-a95s                                                    1   \n",
       "tcga-2y-a9gs                                                    1   \n",
       "tcga-2y-a9gt                                                    1   \n",
       "tcga-2y-a9gu                                                    1   \n",
       "tcga-2y-a9gv                                                    1   \n",
       "\n",
       "                             patient.samples.sample.shortest_dimension  \\\n",
       "patient.bcr_patient_barcode                                              \n",
       "tcga-2v-a95s                                                       NaN   \n",
       "tcga-2y-a9gs                                                       NaN   \n",
       "tcga-2y-a9gt                                                       NaN   \n",
       "tcga-2y-a9gu                                                       NaN   \n",
       "tcga-2y-a9gv                                                       NaN   \n",
       "\n",
       "                            patient.samples.sample.time_between_clamping_and_freezing  \\\n",
       "patient.bcr_patient_barcode                                                             \n",
       "tcga-2v-a95s                                                               NaN          \n",
       "tcga-2y-a9gs                                                               NaN          \n",
       "tcga-2y-a9gt                                                               NaN          \n",
       "tcga-2y-a9gu                                                               NaN          \n",
       "tcga-2y-a9gv                                                               NaN          \n",
       "\n",
       "                             patient.samples.sample.time_between_excision_and_freezing  \\\n",
       "patient.bcr_patient_barcode                                                              \n",
       "tcga-2v-a95s                                                               NaN           \n",
       "tcga-2y-a9gs                                                               NaN           \n",
       "tcga-2y-a9gt                                                               NaN           \n",
       "tcga-2y-a9gu                                                               NaN           \n",
       "tcga-2y-a9gv                                                               NaN           \n",
       "\n",
       "                            patient.samples.sample.tissue_type  \\\n",
       "patient.bcr_patient_barcode                                      \n",
       "tcga-2v-a95s                                               NaN   \n",
       "tcga-2y-a9gs                                               NaN   \n",
       "tcga-2y-a9gt                                               NaN   \n",
       "tcga-2y-a9gu                                               NaN   \n",
       "tcga-2y-a9gv                                               NaN   \n",
       "\n",
       "                            patient.samples.sample.tumor_descriptor  \\\n",
       "patient.bcr_patient_barcode                                           \n",
       "tcga-2v-a95s                                                    NaN   \n",
       "tcga-2y-a9gs                                                    NaN   \n",
       "tcga-2y-a9gt                                                    NaN   \n",
       "tcga-2y-a9gu                                                    NaN   \n",
       "tcga-2y-a9gv                                                    NaN   \n",
       "\n",
       "                             patient.samples.sample.tumor_pathology  \\\n",
       "patient.bcr_patient_barcode                                           \n",
       "tcga-2v-a95s                                                    NaN   \n",
       "tcga-2y-a9gs                                                    NaN   \n",
       "tcga-2y-a9gt                                                    NaN   \n",
       "tcga-2y-a9gu                                                    NaN   \n",
       "tcga-2y-a9gv                                                    NaN   \n",
       "\n",
       "                             patient.samples.sample.vial_number  \n",
       "patient.bcr_patient_barcode                                      \n",
       "tcga-2v-a95s                                                  a  \n",
       "tcga-2y-a9gs                                                  a  \n",
       "tcga-2y-a9gt                                                  a  \n",
       "tcga-2y-a9gu                                                  a  \n",
       "tcga-2y-a9gv                                                  a  \n",
       "\n",
       "[5 rows x 1586 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "clinical = pd.read_csv(clinical_file,compression=\"gzip\").set_index('patient.bcr_patient_barcode')\n",
    "display(clinical.shape)\n",
    "display(clinical.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Gene level distribution\n",
    "\n",
    "In this section, we will investigate the value distribution of genes in our dataset.\n",
    "\n",
    "<br>\n",
    "\n",
    "#### Sample Questions:\n",
    "\n",
    "- What is the range of values for a given gene?\n",
    "\n",
    "- What is the distribution of values for a given gene?\n",
    "\n",
    "- Are there higher than average or lower than average expression of genes?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Dimension reduction based on gene expression\n",
    "\n",
    "If we were working for a smaller dataset we'd be able to plot the few genes we had on a expression by sample graph. However, we are working with 20k genes and will need a better method than creating 20k separate plots. One way to visualize sample distances is to reduce the dimensional space of gene expression across samples.\n",
    "\n",
    "Using principal component analysis (PCA) we can project the data points into a 2D plane so that each axis or component captures the greatest variation of the sample expression. \n",
    "\n",
    "PCA can be useful to characterize the distances between samples and identify groups of samplings that may be farther or closer apart. We can also identify any samples that might be worth excluding from our analysis. Documentation on sklearn's PCA function can be found [here](https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.PCA.html)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from sklearn.decomposition import PCA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pca = PCA(n_components=2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Differential Expression Analysis\n",
    "\n",
    "We can statistically test for a difference in gene expression by performing a hypothesis test for each gene to see whether there is evidence to decide that expression is signficantly different between conditions. \n",
    "In this section, we will investigate differential expression results derived from the [DESeq2] package in R. Also see this [vignette](https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html) on how to do these analyses, as well as understand the methods. \n",
    "\n",
    "Briefly, a Sample x Gene matrix of raw count reads and a matching matrix of phenotypes (eg, treatment or condition) for each sample are inputted into DEseq2. In order to account for RNA composition bias and library size, when only a small number of genes are highly expressed in one condition vs another, these genes are normalized by diving the expression by the mean expression for each gene across samples. \n",
    "\n",
    "DEseq2 fits a [negative binomial](https://en.wikipedia.org/wiki/Negative_binomial_distribution) generalized linear model to each gene and uses the [Wald test](https://www.statisticshowto.datasciencecentral.com/wald-test) for significance. Outliers are detected using Cooke's distance and removed from the dataset. Genes with low normalized mean expression values below a threshbold are also removed to improve detection power. https://chipster.csc.fi/manual/deseq2.html\n",
    "\n",
    "#### Sample questions:\n",
    "- Which genes are differentially expressed? Are they positively or negatively expressed compared to your control?\n",
    "- What do these genes do? Which pathways are they involved in?\n",
    "- Are there related clinical phenotypes which might show similar differences in expression?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "### Running differential expression analysis via DEseq2\n",
    "We've created a subprocess to run the R package DEseq2 with our data and return the results in a python-friendly format. \n",
    "\n",
    "To run this script on the command line use:\n",
    "```\n",
    "Rscript TCGA_differential_expression_analysis.R \"\" lihc_rnaseq.csv.gz lihc_clinical.csv.gz 100 patient.gender female male\n",
    "```\n",
    "\n",
    "(script name | rna_seq_file.csv.gz | clinical_data_file.csv.gz | num_genes_to_run | clinical_attribute_col | atttribute 1 | attribute 2\n",
    "\n",
    "Use the output of this file (lihc_DESeq2_100_sampled_genes_patient.gender_female_vs_male.csv) to plot and identify differentially-expressed genes. You can change the clinical columns, attributes and number of genes to explore additional clinical variables."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>stat</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "      <th>Gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>0.386931</td>\n",
       "      <td>-0.460346</td>\n",
       "      <td>0.498665</td>\n",
       "      <td>-0.923158</td>\n",
       "      <td>3.559248e-01</td>\n",
       "      <td>5.813439e-01</td>\n",
       "      <td>TRPV5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>0.022940</td>\n",
       "      <td>-0.073650</td>\n",
       "      <td>2.238344</td>\n",
       "      <td>-0.032904</td>\n",
       "      <td>9.737514e-01</td>\n",
       "      <td>9.994934e-01</td>\n",
       "      <td>KRTAP10.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>13.638057</td>\n",
       "      <td>1.692997</td>\n",
       "      <td>0.210093</td>\n",
       "      <td>8.058310</td>\n",
       "      <td>7.735626e-16</td>\n",
       "      <td>7.580913e-14</td>\n",
       "      <td>ATP6V1B1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>175.257291</td>\n",
       "      <td>0.090832</td>\n",
       "      <td>0.129924</td>\n",
       "      <td>0.699115</td>\n",
       "      <td>4.844801e-01</td>\n",
       "      <td>6.718166e-01</td>\n",
       "      <td>SARM1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>235.103073</td>\n",
       "      <td>0.004816</td>\n",
       "      <td>0.151323</td>\n",
       "      <td>0.031824</td>\n",
       "      <td>9.746128e-01</td>\n",
       "      <td>9.994934e-01</td>\n",
       "      <td>KCNMA1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     baseMean  log2FoldChange     lfcSE      stat        pvalue          padj  \\\n",
       "0    0.386931       -0.460346  0.498665 -0.923158  3.559248e-01  5.813439e-01   \n",
       "1    0.022940       -0.073650  2.238344 -0.032904  9.737514e-01  9.994934e-01   \n",
       "2   13.638057        1.692997  0.210093  8.058310  7.735626e-16  7.580913e-14   \n",
       "3  175.257291        0.090832  0.129924  0.699115  4.844801e-01  6.718166e-01   \n",
       "4  235.103073        0.004816  0.151323  0.031824  9.746128e-01  9.994934e-01   \n",
       "\n",
       "        Gene  \n",
       "0      TRPV5  \n",
       "1  KRTAP10.7  \n",
       "2   ATP6V1B1  \n",
       "3      SARM1  \n",
       "4     KCNMA1  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "de_gender = pd.read_csv('lihc_DESeq2_100_sampled_genes_patient.gender_female_vs_male.csv')\n",
    "de_gender.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create a diagnostic plots to help us visualize the data. An MA-plot shows the log2 fold changes from the treatment over the mean of normalized counts, i.e. the average of counts normalized by size factor. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Pathway analysis\n",
    "We can do a brief survey of what kinds of pathways and functions some of our differentially expressed genes are involved in. Using the [BioMart](http://www.biomart.org/martservice_9.html) service to annotate our genes with gene ontology (GO) terms."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from pybiomart import Server, Dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GO term accession</th>\n",
       "      <th>GO term name</th>\n",
       "      <th>GO term definition</th>\n",
       "      <th>HGNC symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RNA5SP65</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RN7SKP195</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RNU7-29P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SNORA77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MIR4425</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>63320</td>\n",
       "      <td>GO:0030182</td>\n",
       "      <td>neuron differentiation</td>\n",
       "      <td>The process in which a relatively unspecialize...</td>\n",
       "      <td>NIF3L1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>63321</td>\n",
       "      <td>GO:1903507</td>\n",
       "      <td>negative regulation of nucleic acid-templated ...</td>\n",
       "      <td>Any process that stops, prevents or reduces th...</td>\n",
       "      <td>NIF3L1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>63322</td>\n",
       "      <td>GO:0005739</td>\n",
       "      <td>mitochondrion</td>\n",
       "      <td>A semiautonomous, self replicating organelle t...</td>\n",
       "      <td>NIF3L1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>63323</td>\n",
       "      <td>GO:0042802</td>\n",
       "      <td>identical protein binding</td>\n",
       "      <td>Interacting selectively and non-covalently wit...</td>\n",
       "      <td>NIF3L1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>63324</td>\n",
       "      <td>GO:0008134</td>\n",
       "      <td>transcription factor binding</td>\n",
       "      <td>Interacting selectively and non-covalently wit...</td>\n",
       "      <td>NIF3L1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>63325 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      GO term accession                                       GO term name  \\\n",
       "0                   NaN                                                NaN   \n",
       "1                   NaN                                                NaN   \n",
       "2                   NaN                                                NaN   \n",
       "3                   NaN                                                NaN   \n",
       "4                   NaN                                                NaN   \n",
       "...                 ...                                                ...   \n",
       "63320        GO:0030182                             neuron differentiation   \n",
       "63321        GO:1903507  negative regulation of nucleic acid-templated ...   \n",
       "63322        GO:0005739                                      mitochondrion   \n",
       "63323        GO:0042802                          identical protein binding   \n",
       "63324        GO:0008134                       transcription factor binding   \n",
       "\n",
       "                                      GO term definition HGNC symbol  \n",
       "0                                                    NaN    RNA5SP65  \n",
       "1                                                    NaN   RN7SKP195  \n",
       "2                                                    NaN    RNU7-29P  \n",
       "3                                                    NaN     SNORA77  \n",
       "4                                                    NaN     MIR4425  \n",
       "...                                                  ...         ...  \n",
       "63320  The process in which a relatively unspecialize...      NIF3L1  \n",
       "63321  Any process that stops, prevents or reduces th...      NIF3L1  \n",
       "63322  A semiautonomous, self replicating organelle t...      NIF3L1  \n",
       "63323  Interacting selectively and non-covalently wit...      NIF3L1  \n",
       "63324  Interacting selectively and non-covalently wit...      NIF3L1  \n",
       "\n",
       "[63325 rows x 4 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset = Dataset(name='hsapiens_gene_ensembl',\n",
    "                  host='http://www.ensembl.org')\n",
    "\n",
    "attributes = [\n",
    "    'go_id',\n",
    "    'name_1006',\n",
    "    'definition_1006',\n",
    "    'hgnc_symbol',\n",
    "]\n",
    "\n",
    "bm = dataset.query(attributes=attributes,\n",
    "              filters={'chromosome_name': ['1','2']})\n",
    "\n",
    "bm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Clinical data type investigation\n",
    "\n",
    "Now that we've gone through the steps of exploratory analysis for a single clinical attribute, you can repeat this process and explore how other clinical factors may change gene expression. In this section, we will investigate the diversity of the clinical data. Explore the clinical attributes available in the dataset. \n",
    "\n",
    "We are interested in understanding whether there are RNA signatures that are change according to disease progression for hepatocellular carcinoma. Are there any relevant attributes in the clinical data which are relevant to identifying this find of biomarker? \n",
    "\n",
    "\n",
    "#### Sample Questions:\n",
    "\n",
    "- How many unique values are there for a given clinical attribute?\n",
    "\n",
    "- How can we define an appropriate response variable for supervised learning?\n",
    " \n",
    "- What clinical attributes can be used to identify biomarkers for disease progression?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Set up for supervised learning \n",
    "\n",
    "In this section, we will set up a supervised learning paradigm using the Genes within the RNASeq data as predictors and a clinical attribute as a response variable. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "env-tcga",
   "language": "python",
   "name": "env-tcga"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
